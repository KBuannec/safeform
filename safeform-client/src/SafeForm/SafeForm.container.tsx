import SafeForm from './SafeForm';
import { useSafeForm } from './SafeForm.hook';

const SafeFormContainer = () => {
    const { form, errors, success, handleChange, handleSubmit } = useSafeForm();

    return (
        <SafeForm
            form={form}
            errors={errors}
            success={success}
            onChange={handleChange}
            onSubmit={handleSubmit}
        />
    );
};

export default SafeFormContainer;
