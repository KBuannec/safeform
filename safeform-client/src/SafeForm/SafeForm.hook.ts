import { useState } from 'react';
import { z } from 'zod';

const formSchema = z.object({
    name: z.string().min(1, 'Nom requis'),
    email: z.string().email('Email invalide')
});

export type FormData = z.infer<typeof formSchema>;

export const useSafeForm = () => {
    const [form, setForm] = useState<FormData>({ name: '', email: '' });
    const [errors, setErrors] = useState<Partial<Record<keyof FormData, string>>>({});
    const [success, setSuccess] = useState<string | null>(null);

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setForm({ ...form, [e.target.name]: e.target.value });
    };

    const handleSubmit = async (e: React.FormEvent) => {
        e.preventDefault();

        const result = formSchema.safeParse(form);
        if (!result.success) {
            const fieldErrors = result.error.flatten().fieldErrors;
            setErrors({
                name: fieldErrors.name?.[0],
                email: fieldErrors.email?.[0],
            });
            return; // ✅ on bloque l'envoi si les données sont invalides
        }

        setErrors({});
        try {
            const res = await fetch('http://localhost:3000/submit', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(form)
            });

            if (res.ok) {
                setSuccess('Formulaire envoyé avec succès !');
                setForm({ name: '', email: '' });
            } else {
                setSuccess("Erreur lors de l'envoi");
            }
        } catch (error) {
            console.error(error);
            setSuccess("Une erreur est survenue");
        }
    };


    return {
        form,
        errors,
        success,
        handleChange,
        handleSubmit
    };
};
