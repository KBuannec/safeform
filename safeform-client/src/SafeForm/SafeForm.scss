$primary-color: #2563eb;
$danger-color: #dc2626;
$success-color: #16a34a;

.safeform {
  background: white;
  padding: 2rem;
  border-radius: 0.75rem;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
  width: 100%;
  max-width: 500px;

  label {
    display: block;
    margin-bottom: 0.5rem;
    color: #333;
    font-weight: 500;
  }

  input {
    width: 100%;
    padding: 0.6rem;
    margin-bottom: 1rem;
    border: 1px solid #ccc;
    border-radius: 0.375rem;
    font-size: 1rem;
  }

  button {
    width: 100%;
    background-color: $primary-color;
    color: white;
    padding: 0.75rem;
    border: none;
    border-radius: 0.375rem;
    font-weight: bold;
    cursor: pointer;
    transition: background-color 0.3s ease;

    &:hover {
      background-color: #1d4ed8;
    }
  }

  .error {
    color: $danger-color;
    font-size: 0.875rem;
  }

  .success {
    color: $success-color;
    text-align: center;
    font-size: 0.875rem;
  }
}

.safeform-wrapper {
  position: fixed;
  top: 0;
  left: 0;
  height: 100vh;
  width: 100vw;

  display: flex;
  justify-content: center;
  align-items: center;

  background-color: #f9fafb;
  padding: 2rem;
  box-sizing: border-box;
}

