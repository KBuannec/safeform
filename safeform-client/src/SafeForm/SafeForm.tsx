import './SafeForm.scss';

type Props = {
    form: { name: string; email: string };
    errors: { name?: string; email?: string };
    success: string | null;
    onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
    onSubmit: (e: React.FormEvent) => void;
};

const SafeForm = ({ form, errors, success, onChange, onSubmit }: Props) => {
    return (
        <form onSubmit={onSubmit} className="safeform">
            <h2>üõ°Ô∏è SafeForm</h2>

            <label>Nom</label>
            <input
                type="text"
                name="name"
                value={form.name}
                onChange={onChange}
            />
            {errors.name && <p className="error">{errors.name}</p>}

            <label>Email</label>
            <input
                type="email"
                name="email"
                value={form.email}
                onChange={onChange}
            />
            {errors.email && <p className="error">{errors.email}</p>}

            <button type="submit">Envoyer</button>

            {success && <p className="success">{success}</p>}
        </form>

    );
};

export default SafeForm;
