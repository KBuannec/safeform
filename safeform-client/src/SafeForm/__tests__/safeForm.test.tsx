import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import SafeForm from '../SafeForm';

describe('SafeForm component', () => {
    const baseProps = {
        form: { name: '', email: '' },
        errors: {},
        success: null,
        onChange: vi.fn(),
        onSubmit: vi.fn(),
    };

    it('affiche le titre et les champs', () => {
        render(<SafeForm {...baseProps} />);

        expect(screen.getByRole('heading', { name: /safeform/i })).toBeInTheDocument();
        expect(screen.getByLabelText(/nom/i)).toBeInTheDocument();
        expect(screen.getByLabelText(/email/i)).toBeInTheDocument();
    });

    it("affiche les messages d'erreur", () => {
        render(
            <SafeForm
                {...baseProps}
                errors={{ name: 'Nom requis', email: 'Email invalide' }}
            />
        );

        expect(screen.getByText('Nom requis')).toBeInTheDocument();
        expect(screen.getByText('Email invalide')).toBeInTheDocument();
    });

    it('appelle onChange au changement de champ', () => {
        render(<SafeForm {...baseProps} />);

        fireEvent.change(screen.getByLabelText(/nom/i), {
            target: { value: 'Jean', name: 'name' },
        });
        expect(baseProps.onChange).toHaveBeenCalled();

        fireEvent.change(screen.getByLabelText(/email/i), {
            target: { value: 'jean@example.com', name: 'email' },
        });
        expect(baseProps.onChange).toHaveBeenCalled();
    });

    it('appelle onSubmit lors de la soumission du formulaire', () => {
        render(<SafeForm {...baseProps} />);

        fireEvent.submit(screen.getByText('Envoyer').closest('form')!);
        expect(baseProps.onSubmit).toHaveBeenCalled();
    });

    it('correspond au snapshot', () => {
        const { asFragment } = render(<SafeForm {...baseProps} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
