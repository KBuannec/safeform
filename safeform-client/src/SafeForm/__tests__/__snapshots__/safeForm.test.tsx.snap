// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SafeForm component > correspond au snapshot 1`] = `
<DocumentFragment>
  <form
    class="safeform"
  >
    <h2>
      🛡️ SafeForm
    </h2>
    <label
      for="name"
    >
      Nom
    </label>
    <input
      id="name"
      name="name"
      type="text"
      value=""
    />
    <label
      for="email"
    >
      Email
    </label>
    <input
      id="email"
      name="email"
      type="email"
      value=""
    />
    <button
      type="submit"
    >
      Envoyer
    </button>
  </form>
</DocumentFragment>
`;
